packages/vm/src/runBlock.ts

```diff
@@ -208,66 +215,89 @@ export async function runBlock(this: VM, opts: RunBlockOpts): Promise<RunBlockRe
     const gasUsed = result.gasUsed
     const receiptTrie = result.receiptsRoot
     const transactionsTrie = await _genTxTrie(block)
-    const generatedFields = { stateRoot, bloom, gasUsed, receiptTrie, transactionsTrie }
+    const generatedFields = {
+      stateRoot,
+      bloom,
+      gasUsed,
+      receiptTrie,
+      transactionsTrie,
+      requestsRoot,
+    }
     const blockData = {
       ...block,
       header: { ...block.header, ...generatedFields },
     }
     block = Block.fromBlockData(blockData, { common: this.common })
-  } else if (!this.common.isActivatedEIP(6800)) {
-    // Only validate the following headers if verkle blocks aren't activated
-    if (equalsBytes(result.receiptsRoot, block.header.receiptTrie) === false) {
-      if (this.DEBUG) {
-        debug(
-          `Invalid receiptTrie received=${bytesToHex(result.receiptsRoot)} expected=${bytesToHex(
-            block.header.receiptTrie
-          )}`
-        )
+  } else {
+    if (this.common.isActivatedEIP(7685)) {
+      const valid = await block.requestsTrieIsValid()
```

Comment:
Shouldn't we do this check before running all txs?

Reply:
If the `generate` option is called, the `requestsRoot` passed in the `block` will be disregarded and the correct one derived so checking before we validate the `generate` flag doesn't make sense.  We validate it here if we are re-executing a block to ascertain its validity.

Reply:
If `generate` is called we generate the `requestsRoot` and put that in the block (this is correct). However, this check is done in case we are not using `generate` (so `generate` is `false`). The `block.requestsTrieIsValid` is called on whatever block is provided - this itself, should not change (so this can be done before running any txs). However what you also want to check is that the `block.genRequestsTrieRoot(requests)` (so not `block.requests`, but the `requests` we just have accumulated in `accumulateRequests(this)`) equals the provided `block.header.requestsRoot`. If this is not the case, then the block thus has a different requests root and the block is incorrect.

Reply:
Ah, that makes sense.  Will push an update shortly.